pandas/core/groupby/generic.py

```diff
@@ -485,8 +485,13 @@ def transform(self, func, *args, engine="cython", engine_kwargs=None, **kwargs):
         # If func is a reduction, we need to broadcast the
         # result to the whole group. Compute func result
         # and deal with possible broadcasting below.
+        # Temporarily set observed for dealing with
+        # categoricals so we don't have to convert dtypes.
+        observed = self.observed
```

Comment:
woa? can we not simply handle this at a lower level?

Reply:
- Computing with `observed=True` is much more performant than `observed=False` when the category combinations are not saturated, and slightly less performant when they are. So it seems best to me to run with `observed=True` since transform will disregard any unobserved combinations.
- Computing with `observed=True` makes it so that we don't have to do a type conversion float to int because of `np.nan` for missing combinations.
- The next step is to call `getattr(self, func)(*args, **kwargs)`. So I don't see how to handle this at a lower level.

That said, the code as-is is definitely not good - `self.observed` should be modified using a context manager in case the reduction fails. Something like:

    with temp_setattr(self, observed=True) as obj:
        getattr(obj, func)(*args, **kwargs)

where `temp_setattr` is a context manager like the one [found here](http://code.activestate.com/recipes/577089-context-manager-to-temporarily-set-an-attribute-on/). Not sure if this is a more palatable change.

Reply:
what i mean is, why don't you just pass `observed=True` directly in the function?

Reply:
I think you're suggesting doing something like `getattr(self, func)(*args, **kwargs, observed=True)`. Assuming args and kwargs are empty, that is akin to `df.groupby(keys).sum(observed=True)`; but `sum` doesn't take observed as an argument, only `groupby` does.

Reply:
ok, can you make a context manager for this, like `_group_selection_context` conceptually.